File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/processors/gbd/ServerProducer.java
package ru.kmorozov.library.data.loader.processors.gbd;

import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;
import ru.kmorozov.gbd.core.config.GBDOptions;
import ru.kmorozov.gbd.core.logic.context.IBookListProducer;
import ru.kmorozov.gbd.core.logic.library.LibraryFactory;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

@Component
public class ServerProducer implements IBookListProducer {

    private Set<String> ids;

    @Value("${library.gbd.ids}")
    public String defaultIds;

    @Autowired
    @Lazy
    private DbContextLoader dbCtx;

    @Override
    public Set<String> getBookIds() {
        String bookId = GBDOptions.getBookId();

        if (!StringUtils.isEmpty(bookId) && LibraryFactory.isValidId(bookId))
            this.ids = new HashSet<>(Collections.singletonList(bookId));

        if (this.ids == null)
            if (StringUtils.isEmpty(this.defaultIds))
                this.ids = this.dbCtx.getBookIdsList();
            else
                this.ids = new HashSet(Arrays.asList(this.defaultIds.split(",")));

        return this.ids;
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/processors/gbd/ServerProducer.kt
package ru.kmorozov.library.data.loader.processors.gbd

import com.google.common.base.Strings
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.context.IBookListProducer
import ru.kmorozov.gbd.core.logic.library.LibraryFactory
import java.util.*

@Component
open class ServerProducer : IBookListProducer {

    private var ids: Set<String>? = null

    @Value("\${library.gbd.ids}")
    var defaultIds: String? = null

    @Autowired
    @Lazy
    private val dbCtx: DbContextLoader? = null

    override val bookIds: Set<String>
        get() {
            val bookId = GBDOptions.bookId

            if (!Strings.isNullOrEmpty(bookId) && LibraryFactory.isValidId(bookId))
                ids = HashSet(listOf(bookId))

            if (ids == null)
                if (Strings.isNullOrEmpty(defaultIds))
                    ids = dbCtx!!.bookIdsList
                else
                    ids = HashSet(Arrays.asList(*defaultIds!!.split(",".toRegex()).dropLastWhile { it.isEmpty }.toTypedArray()))

            return ids!!
        }
}
